src/sandstorm/indexer.capnp

```diff
@@ -0,0 +1,139 @@
+# Sandstorm - Personal Cloud Sandbox
+# Copyright (c) 2017 Sandstorm Development Group, Inc. and contributors
+# All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+@0xcae98639575b2b35;
+
+$import "/capnp/c++.capnp".namespace("sandstorm");
+
+using Grain = import "grain.capnp";
+
+interface IndexingSession(Metadata) {
+  # This is a UiView session type, created by calling UiView.newSession().
+  #
+  # Sandstorm requests a session of this type when it wants to index a grain for search purposes.
+  #
+  # Indexing sessions, like any other sessions, *must* pay attention to the UserInfo passed to
+  # `newSession()`; only content which is visible to that user can be indexed.
+  #
+  # Note that, as an optimization, Sandstorm may start out assuming that all users see exactly
+  # the same content, and may do all indexing as an anonymous user, perhaps with no permissions.
+  # However, if the app logs an ActivityEvent that specifies that it requires specific permissions
+  # or is visible only to certain users, Sandstorm uses that as a hint to index the path associated
+  # with that event separately specifically for people who can see said event. Hence, if your app
+  # has content that is visible to some users but not others, and that needs to be indexed, it
+  # should be sure to log appropriate activity events.
+
+  indexAll @0 (indexer :GrainIndexer(Metadata));
+  # Asks the app to iterate through the grain's entire text contents, pushing it all to the given
+  # indexer capability.
+
+  indexPaths @1 (paths :List(Text), indexer :GrainIndexer(Metadata));
+  # Index a specific list of paths. If a path in the list doesn't exist, calls `indexer.index()`
+  # with `content` = null for that path.
+}
+
+interface GrainIndexer(Metadata) {
+  # Capability used to index the content of a grain.
+  #
+  # This is a one-way capability. Although GrainIndexer is implemented by the indexer and is called
```

Comment:
I only vaguely understand what you're intending to protect against by making this a "one-way" capability. Should we be worried that capabilities could be smuggled through the `metadata` field of `IndexableContent`?